src/node_main_instance.cc

```diff
@@ -49,6 +53,14 @@ NodeMainInstance::NodeMainInstance(const SnapshotData* snapshot_data,
       NewIsolate(isolate_params_.get(), event_loop, platform, snapshot_data);
   CHECK_NOT_NULL(isolate_);
 
+  cpp_heap_ = CppHeap::Create(
+      platform,
+      CppHeapCreateParams{{},
+                          WrapperDescriptor(BaseObject::kEmbedderType,
```

Comment:
The type and index is meant for identifying embedder (addons in this case) objects, which I think technically are separate from the internal Baseobjects? I think it would be more robust if we make the convention for addon cppgc wrappers separate - for example, simply declare a `constexpr WrapperDescriptor` in `node.h` with these parameters. (I wonder how users can create cppgc-managed objects without access to the embedder id - do they call `SetAlignedPointerInInternalField()` to some pointer that points to a hard-coded copy of that id?)

Reply:
If we are using `BaseObject::kEmbedderType` and `BaseObject::kSlot` here to ensure that the BaseObject layout is not in conflict with this, perhaps we could declare `constexpr WrapperDescriptor` in `node.h`, and internally we `static_assert` that the indices are compatible (matching BaseObject offsets).

Reply:
There can (at least currently) be only one `CppHeap` for a given `isolate`.  All objects that live in that heap need to then all agree upon the ids and slots used here.

If there are two add-ons both using `cppgc` to create objects, they need to be compatible.

The API used to create such an object is:
```cpp
T* obj = cppgc::MakeGarbageCollected<T>(
  isolate->GetCppHeap()->GetAllocationHandle(),
  args...);
```

The `AllocationHandle` encapsulates which `CppHeap` the object is being created in, and therefore which values are used for the ids and slots.  Add-ons do not need to be explicitly aware which ids and slots are being used.

Reply:
About https://github.com/nodejs/node/pull/45704#discussion_r1037419323, IIUC this is something that the user needs to do to create cppgc-managed objects in addons:

```c++
uint16_t wrappable_type = kNodeEmbedderIdForCppgc;
// The lifetime of wrappable_object is managed by V8
auto* wrappable_object =
    cppgc::MakeGarbageCollected<Wrappable>(isolate()->GetCppHeap()->GetAllocationHandle());
Local<Object> object = Object::New(isolate)
object->SetAlignedPointerInInternalField(kNodeEmbedderWrappableTypeIndex,
                                         &wrappable_type);
object->SetAlignedPointerInInternalField(kNodeEmbedderWrappableInstanceIndex,
                                         wrappable_object);
```

This is different from Electron's case because there they are configuring the cpp heap, so they are already aware of the wrapper descriptor. But here we are the one picking the wrapper descriptor, so we should actually expose the ID and the `wrappable_instance_index` that we pick, because addons are supposed to use that index to store their `GarbageCollected` wrappables.

Reply:
I guess it depends what you're doing, but if you want to introduce an edge to this `cppgc` object from a `v8::Object` you can put the pointer into the internal field like that, yes.  We do actually do that.

I guess today people use `node::ObjectWrap`.  Maybe there should be a variant of that for `cppgc` that encapsulates all these gnarly details?

Reply:
I think this model probably is a bit too different from what `ObjectWrap()` is doing now (which still relies on a persistent handle). Perhaps exposing a helper for setting this is already enough? Something like

```
void SetCppgcWrappable(Local<Object> object, void* garbageCollected);
```

which does the snippet above.